src/v/transform/logging/log_manager.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/property.h"
+#include "model/transform.h"
+#include "ssx/semaphore.h"
+#include "transform/logging/event.h"
+#include "transform/logging/io.h"
+#include "wasm/logger.h"
+
+#include <seastar/core/lowres_clock.hh>
+
+#include <absl/container/flat_hash_map.h>
+
+#include <utility>
+
+namespace transform::logging {
+
+namespace detail {
+template<typename ClockType>
+class flusher;
+}
+
+template<typename ClockType = ss::lowres_clock>
+class manager {
+    static_assert(
+      std::is_same_v<ClockType, ss::lowres_clock>
+        || std::is_same_v<ClockType, ss::manual_clock>,
+      "Only lowres or manual clocks are supported");
+
+public:
+    manager() = delete;
+    ~manager();
+    manager(const manager&) = delete;
+    manager& operator=(const manager&) = delete;
+    manager(manager&&) = delete;
+    manager& operator=(manager&&) = delete;
+
+    explicit manager(
+      model::node_id,
+      std::unique_ptr<client>,
+      size_t buffer_cap,
+      config::binding<size_t> line_limit,
+      config::binding<std::chrono::milliseconds> flush_interval);
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    void enqueue_log(
+      ss::log_level lvl, model::transform_name_view, std::string_view message);
+
+private:
+    // TODO(oren): make configurable?
+    static constexpr double lwm_denom = 10;
+    bool check_lwm() const;
+
+    model::node_id _self;
+    std::unique_ptr<client> _client;
+    config::binding<size_t> _line_limit_bytes;
+    size_t _buffer_limit_bytes;
+    ssize_t _buffer_low_water_mark;
+    ssx::semaphore _buffer_sem;
+    config::binding<std::chrono::milliseconds> _flush_interval_ms;
+
+    ss::gate _gate{};
+    ss::abort_source _as{};
+
+    struct buffer_entry {
+        buffer_entry() = delete;
+        explicit buffer_entry(event event, ssx::semaphore_units units)
+          : event(std::move(event))
+          , units(std::move(units)) {}
+        event event;
+        ssx::semaphore_units units;
+    };
+
+    struct string_hash {
+        using is_transparent = void;
+        [[nodiscard]] size_t operator()(std::string_view txt) const {
+            return std::hash<std::string_view>{}(txt);
+        }
+        [[nodiscard]] size_t operator()(const ss::sstring& txt) const {
+            return std::hash<ss::sstring>{}(txt);
+        }
+    };
+    // per @rockwood
+    // TODO(oren): Evaluate (and probably substitute) `chunked_vector` once it
+    // lands
+    using buffer_t = ss::chunked_fifo<buffer_entry>;
+    absl::flat_hash_map<ss::sstring, buffer_t, string_hash, std::equal_to<>>
```

Comment:
nit: Generally we have been using `btree` to be safe, even though realistically we probably won't have that many transforms on a single core to need it, but I don't think the perf difference matters honestly so 🤷

Reply:
ya, was going to bring this up. no strong preference here; just thought I had seen flat_map in various places

Reply:
Contiguous allocations. `flat_hash` and `flat_node` store elements in a single contiguous memory region, we don't let those go over 128KiB, while `btree` uses chunks that stay under our limit.

https://abseil.io/docs/cpp/guides/container gives some nice pictures of the hash map versions

Reply:
>I don't think the perf difference matters

ah, the *negative* perf difference between `flat_hash` and `btree`. I  was aware of the memory layout difference, I thought you were making specifically a perf argument in favor of `btree`. Confused. Anyway yeah, changing.

Reply:
Oh yeah I think in general flat_hash is faster. Sorry for the confusion!

Reply:
no prob. iffy reading comprehension on my part i think. anyway changed to `btree_map`.